[ONNX] Introduce SARIF

BowenBao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #85107
* #85918
* #85898
* #85651
* __->__ #85428

That's the parent issue tracking this and more follow up tasks, so will keep open after this.
This PR introduces the python classes for SARIF object model, along with script for generation.

BowenBao committed:
[ONNX] Introduce SARIF

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Introduce SARIF"




[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Introduce SARIF"




[ghstack-poisoned]

justinchuby commented on tools/onnx/sarif/gen_sarif.sh:
```diff
@@ -0,0 +1,32 @@
+#!/bin/bash
```
I would add a brief comment on usage

BowenBao replied:
Added a few lines to start with

justinchuby commented on torch/onnx/_internal/diagnostics/infra/sarif_om/__init__.py:
```diff
@@ -0,0 +1,107 @@
+# This file was generated by jschema_to_python version 1.2.3.
```
Possible to say "Do not edit" in these files to make it clear?

BowenBao replied:
That will be further hacking the jschema_to_python tool, but why not 🤣 Changes will appear in the next PR then.

thiagocrepaldi reviewed (commented):
What files are reviewable, if any? It seems only files within `tools/onnx/sarif/` as the rest is auto generated

justinchuby reviewed (approved):
LGTM🚀

BowenBao commented:
> What files are reviewable, if any? It seems only files within `tools/onnx/sarif/` as the rest is auto generated

Correct.

AllenTiTaiWang commented on tools/onnx/sarif/gen_sarif.sh:
```diff
@@ -0,0 +1,32 @@
+
+# SARIF version
+SARIF_VERSION=2.1.0
+SARIF_SCHEMA_LINK=https://docs.oasis-open.org/sarif/sarif/v2.1.0/cs01/schemas/sarif-schema-2.1.0.json
```
Would this json/version change very often? Curious about the maintenance work on this.

BowenBao replied:
No, I don't expect it to change. However soon we should add CI checks to ensure any modification to these files are made through code generation.

BowenBao committed:
Update on "[ONNX] Introduce SARIF"


~~NotYetFix: https://github.com/pytorch/pytorch/issues/84659 BowenBao should this issue be closed along with this PR?~~
That's the parent issue tracking this and more follow up tasks, so will keep open after this.

[ghstack-poisoned]

thiagocrepaldi commented on tools/onnx/sarif/gen_sarif.sh:
```diff
@@ -0,0 +1,43 @@
+# This script requires the jschema_to_python package to be installed.
+# To install it, run:
+#   pip install jschema_to_python
+ROOT=$(pwd)/../../..
```
```suggestion
set -e -x
ROOT=$(pwd)/../../..
```
Make sure that CI pipelines (or any other script that calls this one) will fail if the script itself fail

abock replied:
This whole script needs to be `shellcheck`'d. Most notably variables are unquoted which can introduce unexpected globbing and splitting.

* https://gist.github.com/abock/620576895d6ae3ed12c9863ac73d283d

BowenBao replied:
I updated the script in the last PR in this stack. LMK if that works

thiagocrepaldi commented on tools/onnx/sarif/gen_sarif.sh:
```diff
@@ -0,0 +1,43 @@
+sarif_schema_file_path=$tmp_dir/sarif-schema-$SARIF_VERSION.json
+wget -O $sarif_schema_file_path $SARIF_SCHEMA_LINK
+
+# TODO: A private branch of jschema_to_python was used to enable
```
why was the private needed? If anyone else needs to run this, which branch they should use (full path is better)? Is the branch `0.0.1.dev29` from upstream?

BowenBao replied:
Oops this is leaked down to this commit. Only the follow ups that generate code with dataclasses and type annotations need this. A private branch is needed because official one only supports attr, which PyTorch does not. And afaik there is no other tools on the market to generate from jschema to python.

thiagocrepaldi commented on torch/onnx/_internal/diagnostics/infra/sarif_om/__init__.py:
```diff
@@ -0,0 +1,108 @@
+from torch.onnx._internal.diagnostics.infra.sarif_om._web_request import WebRequest
+from torch.onnx._internal.diagnostics.infra.sarif_om._web_response import WebResponse
+
+# flake8: noqa
```
was this intended as the last line or should be the first? Shouldn't it be two spaces after `#`?

BowenBao replied:
Let me know if there is strong opinion first line vs last. I found it is much more efficient for bash to append than to prepend.

thiagocrepaldi replied:
if the tooling is ok, I am fine with it. We can change the .flake8 in a future pr and get rid of this anyway

thiagocrepaldi commented on torch/onnx/_internal/diagnostics/infra/sarif_om/__init__.py:
```diff
@@ -0,0 +1,108 @@
+# DO NOT EDIT! This file was generated by jschema_to_python version 0.0.1.dev29,
+# with extension for dataclasses and type annotation.
+
```
How about moving all generated files to a `generated` folder? That possibly helps in automating "cleaning builds" during development

thiagocrepaldi replied:
I am guessing `sarif_om` stands for `sarif_object_model`, but being more explicit in the naming makes things easier for first timers

BowenBao replied:
How about `sarif` then

thiagocrepaldi commented on tools/onnx/sarif/gen_sarif.sh:
```diff
@@ -0,0 +1,43 @@
+echo "SARIF_VERSION = \""$SARIF_VERSION"\"" > $ROOT/$SARIF_DIR/version.py
+echo "SARIF_SCHEMA_LINK = \""$SARIF_SCHEMA_LINK"\"" >> $ROOT/$SARIF_DIR/version.py
+
+# Hack to have flake8 not complain about generated code.
```
Another way is adding the generated folder at `.flake8`'s `exclude` list to skip the folder altogether

BowenBao replied:
Will need Meta review for that. I can do it in another clean up only PR.

thiagocrepaldi replied:
a cleanup pr is ok given the time constraint

BowenBao committed:
Update on "[ONNX] Introduce SARIF"


~~NotYetFix: https://github.com/pytorch/pytorch/issues/84659 BowenBao should this issue be closed along with this PR?~~
That's the parent issue tracking this and more follow up tasks, so will keep open after this.

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Introduce SARIF"


~~NotYetFix: https://github.com/pytorch/pytorch/issues/84659 BowenBao should this issue be closed along with this PR?~~
That's the parent issue tracking this and more follow up tasks, so will keep open after this.

[ghstack-poisoned]

abock reviewed (changes requested):
Looks good, but would love to see the shell script sanitized per shellcheck. Shell scripst are often the source of random problems down the road (e.g. when someone tries to use it with spaces in their PWD).

BowenBao commented:
@pytorchbot merge -g
